wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift

```diff
@@ -376,29 +376,70 @@ public final class MessageSender: MessageSenderInterface {
+            // set expiration so request can be expired later
+            await context.perform {
+                if message.shouldExpire {
+                    message.setExpirationDate()
+                    self.context.saveOrRollback()
+                }
+            }
+
+            let (payload, response) = try await apiProvider.messageAPI(apiVersion: apiVersion)
+                .sendMLSMessage(
+                    message: encryptedData,
+                    conversationID: conversationID,
+                    expirationDate: await context.perform { message.expirationDate }
+                )
+
+            await context.perform {
+                // handle 201 case failed_to_send
+                // https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/556564601/Use+case+sending+a+message+MLS
+                self.mlsPayloadProcessor.updateFailedRecipients(from: payload, for: message)
+                message.delivered(with: response)
+            }
+        } catch let error as SendMLSMessageFailure {
+            switch error {
+            case .mlsStaleMessage:
+                // We should try to repair the conversation for the `mlsStaleMessage` error.
+                // This error indicates that the message was not encrypted in the latest epoch.
+                let operation: () async throws -> Void = { [weak self] in
+                    try await self?.sendMessage(message: message)
+                }
+
+                try await handleMLSStaleMessageError(
+                    groupID: groupID,
+                    mlsService: mlsService,
+                    operation: operation
+                )
+            default:
+                throw error
             }
         }
+    }
 
-        let (payload, response) = try await apiProvider.messageAPI(apiVersion: apiVersion)
-            .sendMLSMessage(
-                message: encryptedData,
-                conversationID: conversationID,
-                expirationDate: await context.perform { message.expirationDate }
-            )
+    private func handleMLSStaleMessageError(
+        groupID: MLSGroupID,
+        mlsService: MLSServiceInterface,
+        operation: () async throws -> Void
+    ) async throws {
+        do {
+            await mlsService.fetchAndRepairGroup(with: groupID)
+            try await operation()
+        } catch let error as MessageSendError {
+            guard retryCount < maxRetryAttempts else {
+                retryCount = 0
+                throw error
+            }
 
-        await context.perform {
-            // handle 201 case failed_to_send
-            // https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/556564601/Use+case+sending+a+message+MLS
-            self.mlsPayloadProcessor.updateFailedRecipients(from: payload, for: message)
-            message.delivered(with: response)
+            retryCount += 1
+
+            try await operation()
+        } catch {
+            throw error
```

Comment:
suggestion: no need to catch if all it does is throws the error... if you you commit these two lines it will have the same effect.

Reply:
done